mgr/dashboard: fix rgw connect when using ssl

oOHenry opened the pull request:
The dashboard trys to connect via ssl to an ip address instead of a dns name:

Exception:
```
Traceback (most recent call last):
  File "/usr/share/ceph/mgr/dashboard/rest_client.py", line 399, in do_request
    url, headers=request_headers, params=params, auth=self.auth)
  File "/lib/python3.6/site-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/usr/share/ceph/mgr/dashboard/rest_client.py", line 48, in request
    return super(TimeoutRequestsSession, self).request(*args, **kwargs)
  File "/lib/python3.6/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/lib/python3.6/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/lib/python3.6/site-packages/requests/adapters.py", line 514, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: HTTPSConnectionPool(host='10.88.7.67', port=443): Max retries exceeded with url: /admin/metadata/user?myself (Caused by SSLError(CertificateError("hostname '10.88.7.67' doesn't match either of 's3dev-c01rg
w.lan', '*.s3dev-c01rgw.lan', 's3dev-c01rgw01.lan', '*.s3dev-c01rgw01.lan', 's3dev-c01rgw02.lan', '*.s3dev-c01rgw02.lan', 's3dev-c01rgw03.lan', '*.s3dev-c01rgw03.lan'"
,),))

```
Fixes: http://tracker.ceph.com/issues/56970

aaSharma14 commented:
jenkins test dashboard

aaSharma14 commented:
jenkins test windows

nizamial09 commented:
@oOHenry can you please update the commit message with the tracker id when you get an access?

avanthakkar reviewed (approved):
LGTM!

oOHenry commented:
Can someone point someone to unlock my ceph tracker account? I registered ~8 days ago with the username: hhirsch

nizamial09 commented:
> Can someone point someone to unlock my ceph tracker account? I registered ~8 days ago with the username: hhirsch

@djgalloway @epuertat ^

epuertat commented on src/pybind/mgr/dashboard/services/rgw_client.py:
```diff
@@ -85,6 +85,8 @@ def _determine_rgw_addr(daemon_info: Dict[str, Any]) -> RgwDaemon:
     daemon = RgwDaemon()
     daemon.host = _parse_addr(daemon_info['addr'])
```
One-liner, so the `daemon.host` is only set once:
```suggestion
    # Use hostname for SSL, IP address otherwise
    daemon.host = daemon_info['metadata']['hostname'] if daemon.ssl else _parse_addr(daemon_info['addr'])
```
That said, if we move to hostname-based connection, why not doing it also for HTTP connections?
```suggestion
    daemon.host = daemon_info['metadata']['hostname']
```

oOHenry replied:
>  That said, if we move to hostname-based connection, why not doing it also for HTTP connections?

good point, I added your suggestion to the code

epuertat reviewed (commented):
Thanks a lot for this fix, @oOHenry! Let's wait for @djgalloway to activate your Ceph tracker account.

djgalloway commented:
> Thanks a lot for this fix, @oOHenry! Let's wait for @djgalloway to activate your Ceph tracker account.

Sorry for the delay!  Activated.

oOHenry commented:
Thx for activation my account, I opened an issue and fixed the commit message.
@djgalloway I also added your suggestion, to use the hostname every time not only when ssl is enabled.

nizamial09 commented:
jenkins test make check

nizamial09 commented:
@oOHenry Looks like the make check failure is real
```

________________________ RgwClientTest.test_ssl_verify _________________________

self = <dashboard.tests.test_rgw_client.RgwClientTest testMethod=test_ssl_verify>

    def test_ssl_verify(self):
        Settings.RGW_API_SSL_VERIFY = True
>       instance = RgwClient.admin_instance()

tests/test_rgw_client.py:169: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
services/rgw_client.py:366: in admin_instance
    return RgwClient.instance(daemon_name=daemon_name)
services/rgw_client.py:324: in instance
    RgwClient._daemons = _get_daemons()
services/rgw_client.py:67: in _get_daemons
    daemon = _determine_rgw_addr(daemon_map[key])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

daemon_info = {'addr': '192.168.178.3:49774/1534999298', 'metadata': {'frontend_config#0': 'beast port=8000', 'id': 'daemon1', 'realm_name': 'realm1', 'zone_name': 'zone1', ...}}

    def _determine_rgw_addr(daemon_info: Dict[str, Any]) -> RgwDaemon:
        """
        Parse RGW daemon info to determine the configured host (IP address) and port.
        """
        daemon = RgwDaemon()
>       daemon.host = daemon_info['metadata']['hostname']
E       KeyError: 'hostname'

services/rgw_client.py:86: KeyError
```

oOHenry commented:
jenkins test dashboard

oOHenry commented:
jenkins test api

oOHenry commented:
jenkins test windows

oOHenry commented:
@nizamial09 I think I fixed the: `RgwClientTest.test_ssl_verify` test, but the `ceph API tests` and `ceph windows tests` are failing because of another reason.

pereman2 commented:
jenkins test api

nizamial09 commented:
jenkins test api

nizamial09 commented:
API test failure looks unrelated, retriggering again
```

2022-08-12 16:44:04,187.187 INFO:__main__:----------------------------------------------------------------------
2022-08-12 16:44:04,187.187 INFO:__main__:
2022-08-12 16:44:04,187.187 INFO:__main__:Traceback (most recent call last):
2022-08-12 16:44:04,188.188 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/helper.py", line 271, in setUp
2022-08-12 16:44:04,188.188 INFO:__main__:    self.wait_for_health_clear(self.TIMEOUT_HEALTH_CLEAR)
2022-08-12 16:44:04,188.188 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/ceph_test_case.py", line 175, in wait_for_health_clear
2022-08-12 16:44:04,188.188 INFO:__main__:    self.wait_until_true(is_clear, timeout)
2022-08-12 16:44:04,188.188 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/ceph_test_case.py", line 212, in wait_until_true
2022-08-12 16:44:04,188.188 INFO:__main__:    raise TestTimeoutError("Timed out after {0}s and {1} retries".format(elapsed, retry_count))
2022-08-12 16:44:04,188.188 INFO:__main__:tasks.ceph_test_case.TestTimeoutError: Timed out after 60s and 0 retries
2022-08-12 16:44:04,188.188 INFO:__main__:
Cannot find device "ceph-brx"
2022-08-12 16:44:04,207.207 INFO:__main__:
2022-08-12 16:44:04,208.208 INFO:__main__:----------------------------------------------------------------------
2022-08-12 16:44:04,208.208 INFO:__main__:Ran 73 tests in 818.526s
2022-08-12 16:44:04,208.208 INFO:__main__:
2022-08-12 16:44:04,208.208 INFO:__main__:
```

nizamial09 commented:
jenkins test api

dpeckett commented:
An easy workaround would be to first try connecting via the hostname and then fallback to the address. That should resolve both scenarios without a ton of work.

nizamial09 commented:
I raised a PR https://github.com/ceph/ceph/pull/53133 to fix this. I merged this on main. Will do the backports as well! Rgw has a config-option `rgw_dns_name` which can be used to set the dns name. If its set, dashboard will use that dns name to resolve the ip addr.
